[HPR-918] docs: add docs to show diffs between JSON and HCL2 (#12372)
hashicorp/packer: website/content/docs/templates/json_to_hcl.mdx

---
description: |

>   This section highlights the differences between JSON and HCL2 templates, and explains the parallels between the two specifications.

Reviewer: nit: The comma before the and is not need. 
```suggestion
  This section highlights the differences between JSON and HCL2 templates and explains the parallels between the specifications.
```

page_title: JSON to HCL
---

# Packer templates

> Packer uses templates to orchestrate builds for one or multiple images.

Reviewer: ```suggestion
Packer uses templates to orchestrate builds for one or more images.
```

> In JSON, you would declare a series of builders, provisioners and post-processors to build images.
> In HCL2, things are different, as the configuration language allows you to specify builders
> through sources, and weave them in build blocks.

Reviewer: What do you think about mentioning JSON as a legacy template format? 
```suggestion
In JSON, considered legacy, you would declare a series of builders, provisioners,  and post-processors to build images. In HCL2, things are different, as the configuration language allows you to specify builders
through sources and weave them in build blocks.
```

Author: I'm not sure about the "considered legacy" phrasing, but we can highlight the fact that JSON is indeed deprecated here yes.

Reviewer: That's fair. There is a not at the top of the JSON templates document https://developer.hashicorp.com/packer/docs/templates/legacy_json_templates maybe having a link to the JSON page.

This document aims to explain the parallels between the two configuration files, and what you

> should expect when moving a template away from JSON to HCL2.

Reviewer: What do you think about adding a blurb about the `hcl2_upgrade` command for converting JSON to HCL2 templates and linking out the Upgrade tutorial https://developer.hashicorp.com/packer/tutorials/configuration-language/hcl2-upgrade?

installing packages, writing data to the file system, etc.
They can be defined in both JSON or HCL templates.

### JSON

In a JSON template, provisioners are declared at the top-level of the template as an array of sequentially
invoked components to provision the image you are building. They apply to all the builders defined in the template.

```json
{
  "builders": [
    {
      "type": "null",
      "name": "test",
      "communicator": "none"
    }
  ],
  "provisioners": [
    {
      "type": "shell-local",
      "inline": ["echo test"]
    }
  ]
}
```

### HCL2

In HCL2, provisioners are declared in the `build` block, and apply only to the sources invoked as
part of this build block.
This lets you declare multiple build blocks, each with its own set of provisioners, which will
result in multiple different builds, all run in parallel.

```hcl
source "null" "test" {
  communicator = "none"
}

build {
  sources = ["null.test"]

  provisioner "shell-local" {
    inline = ["echo test"]
  }
}
```

## Post-processors

Post-processors are components that are invoked after the image was produced. Their
role is to consume an artifact produced by a builder, or another post-processor, and

> build a new one from that.

Reviewer: It might help to call out that a new artifact is being built. 
```suggestion
build a new artifact from that. 
```

---
description: |
  This section highlights the differences between JSON and HCL2 templates and explains the parallels between the specifications.
page_title: JSON to HCL
---

# Packer templates

Packer uses templates to orchestrate builds for one or more images.
In [legacy JSON templates](/packer/docs/templates/legacy_json_templates), you would declare a series of builders, provisioners and post-processors to build images.
In [HCL2 templates](/packer/docs/templates/hcl_templates), things are different, as the configuration language allows you to specify builders through sources, and weave them in build blocks.

> This document aims to explain the parallels between the two configuration files, and what you
> should expect when moving a template away from JSON to HCL2.

Reviewer: ```suggestion
This document aims to explain the parallels between the two configuration template types, and what you
should expect when moving a template away from JSON to HCL2.
```

> In addition to this document, you may find the `packer hcl2_upgrade` command useful to help you convert your existing JSON templates to HCL2.
> See the [hcl2_upgrade](/packer/docs/commands/hcl2_upgrade) page for more information on its usage.

Reviewer: ```suggestion
In addition to this document, you may find the `packer hcl2_upgrade` command useful when converting your existing JSON templates to HCL2.
Refer to the [hcl2_upgrade](/packer/docs/commands/hcl2_upgrade) page for more information on its usage.
```

## Builders

Builders are components that are specialized for a platform. Their role is to set the stage for

> the steps that will let you build the final image that you want to build with Packer.

Author: I'm not sure about the `build the final image artifact` phrasing, since I would assume users may see artifacts as a side-effect of building their images, which is essentially what they use Packer for.

### JSON

In JSON, the `builders` attribute declares how a builder will be invoked to build an image.

```json
{
  "builders": [
    {
      "type": "null",
      "name": "test",
      "communicator": "none"
    }
  ]
}
```

### HCL2

> In HCL2, builders are declared through a `source` block. Sources on their own are not enough, and must then
> be invoked in a `build` block.

Reviewer: ```suggestion
In HCL2, builders are declared through a `source` block. Sources on their own are not enough, and must be invoked in a `build` block.
```

> The `build` block serves as a container for all the steps (i.e. `source`, `provisioners` and `post-processors`) to go from a blank image to a finished one.

Reviewer: Nice 💯

```hcl
source "null" "test" {
  communicator = "none"
}

build {
  sources = ["null.test"]
}
```

## Provisioners

> Provisioners are components that modify the state of the machine image you are building, by
> installing packages, writing data to the file system, etc.
> They can be defined in both JSON or HCL templates.

Reviewer: ```suggestion
Provisioners are components that modify the state of the machine image you are building. Provisioners
can be used for installing packages, writing data to the remote file system, or executing remote commands. 
They can be defined in both JSON or HCL templates.
```

### JSON

In a JSON template, provisioners are declared at the top-level of the template as an array of sequentially
invoked components to provision the image you are building. They apply to all the builders defined in the template.

```json
{
  "builders": [
    {
      "type": "null",
      "name": "test",
      "communicator": "none"
    }
  ],
  "provisioners": [
    {
      "type": "shell-local",
      "inline": ["echo test"]
    }
  ]
}
```

### HCL2

In HCL2, provisioners are declared in the `build` block, and apply only to the sources invoked as
part of this build block.
This lets you declare multiple build blocks, each with its own set of provisioners, which will

> result in multiple different builds, all run in parallel.

Reviewer: ```suggestion
result in multiple builds executed in parallel.
```
